tricycle_controller/src/tricycle_controller.cpp

```diff
@@ -0,0 +1,629 @@
+
+  // TODO: clear handles
+  // traction_joint_.clear();
+  // steering_joint_.clear();
+
+  subscriber_is_active_ = false;
+  velocity_command_subscriber_.reset();
+  velocity_command_unstamped_subscriber_.reset();
+
+  received_velocity_msg_ptr_.set(nullptr);
+  is_halted = false;
+  return true;
+}
+
+CallbackReturn TricycleController::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  return CallbackReturn::SUCCESS;
+}
+
+void TricycleController::halt() { traction_joint_[0].velocity_command.get().set_value(0.0); }
+
+CallbackReturn TricycleController::get_traction(
+  const std::string & traction_joint_name, std::vector<TractionHandle> & joint)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Get Wheel Joint Instance");
+
+  // Lookup the velocity state interface
+  const auto state_handle = std::find_if(
+    state_interfaces_.cbegin(), state_interfaces_.cend(),
+    [&traction_joint_name](const auto & interface)
+    {
+      return interface.get_name() == traction_joint_name &&
+             interface.get_interface_name() == HW_IF_VELOCITY;
+    });
+  if (state_handle == state_interfaces_.cend())
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "Unable to obtain joint state handle for %s",
+      traction_joint_name.c_str());
+    return CallbackReturn::ERROR;
+  }
+
+  // Lookup the velocity command interface
+  const auto command_handle = std::find_if(
+    command_interfaces_.begin(), command_interfaces_.end(),
+    [&traction_joint_name](const hardware_interface::LoanedCommandInterface & interface)
+    {
+      return interface.get_name() == traction_joint_name &&
+             interface.get_interface_name() == HW_IF_VELOCITY;
+    });
+  if (command_handle == command_interfaces_.end())
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "Unable to obtain joint state handle for %s",
+      traction_joint_name.c_str());
+    return CallbackReturn::ERROR;
+  }
+
+  // Create the traction joint instance
+  joint.emplace_back(TractionHandle{std::ref(*state_handle), std::ref(*command_handle)});
+  return CallbackReturn::SUCCESS;
+}
+
+CallbackReturn TricycleController::get_steering(
+  const std::string & steering_joint_name, std::vector<SteeringHandle> & joint)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Get Steering Joint Instance");
+
+  // Lookup the velocity state interface
+  const auto state_handle = std::find_if(
+    state_interfaces_.cbegin(), state_interfaces_.cend(),
+    [&steering_joint_name](const auto & interface)
+    {
+      return interface.get_name() == steering_joint_name &&
+             interface.get_interface_name() == HW_IF_POSITION;
+    });
+  if (state_handle == state_interfaces_.cend())
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(), "Unable to obtain joint state handle for %s",
+      steering_joint_name.c_str());
+    return CallbackReturn::ERROR;
+  }
+
+  // Lookup the velocity command interface
+  const auto command_handle = std::find_if(
```

Comment:
this pattern repeats quite often. Could a helper function be of use here? Params would be the interfaces vector, name to look for, interface name (type) to look for -> returns std::optional< handle >

Reply:
Is this important to you? I tried implementing the function you proposed, which is a good idea for reducing lines of codes, I just couldn't really make it compile, for reasons I don't understand. If it's important, I can spend some more time on it.

Reply:
It could be a follow-up issue to refactor this. I think other controllers could benefit from it too. The issue is probably the std::reference_wrapper not playing ball with other ref-style wrappers :) Please open an issue and reference it here, we can close the discussion then

Reply:
I would very much like it in a follow-up issue, here it is: https://github.com/ros-controls/ros2_controllers/issues/388